lib/ecto/adapters/postgres.ex

```diff
@@ -225,28 +243,87 @@ defmodule Ecto.Adapters.Postgres do
 
   @impl true
   def lock_for_migrations(meta, opts, fun) do
-    %{opts: adapter_opts} = meta
+    %{opts: adapter_opts, repo: repo} = meta
 
     if Keyword.fetch(adapter_opts, :pool_size) == {:ok, 1} do
       Ecto.Adapters.SQL.raise_migration_pool_size_error()
     end
 
     opts = Keyword.merge(opts, [timeout: :infinity, telemetry_options: [schema_migration: true]])
-
-    {:ok, result} =
-      transaction(meta, opts, fn ->
-        # SHARE UPDATE EXCLUSIVE MODE is the first lock that locks
-        # itself but still allows updates to happen, see
-        # # https://www.postgresql.org/docs/9.4/explicit-locking.html
-        source = Keyword.get(opts, :migration_source, "schema_migrations")
-        table = if prefix = opts[:prefix], do: ~s|"#{prefix}"."#{source}"|, else: ~s|"#{source}"|
-        {:ok, _} = Ecto.Adapters.SQL.query(meta, "LOCK TABLE #{table} IN SHARE UPDATE EXCLUSIVE MODE", [], opts)
-        fun.()
-      end)
+    lock_strategy = Keyword.get(repo.config(), :migration_lock)
+    {:ok, result} = do_lock_for_migrations(lock_strategy, meta, opts, fun)
 
     result
   end
 
+  defp do_lock_for_migrations(:pg_advisory_lock, meta, opts, fun) do
+    lock = :erlang.phash2("ecto_#{inspect(meta.repo)}")
+    config = meta.repo.config()
+
+    opts =
+      opts
+      |> Keyword.put_new(:migration_advisory_lock_retry_interval_ms, config[:migration_advisory_lock_retry_interval_ms] || 5000)
+      |> Keyword.put_new(:migration_advisory_lock_max_tries, config[:migration_advisory_lock_max_tries] || 100)
+      |> Keyword.put(:migration_advisory_lock_tries, 0)
+
+    advisory_lock(meta, opts, lock, fn ->
+      {:ok, fun.()}
+    end)
+  end
+
+  defp do_lock_for_migrations(_lock_strategy, meta, opts, fun) do
+    transaction(meta, opts, fn ->
+      # SHARE UPDATE EXCLUSIVE MODE is the first lock that locks
+      # itself but still allows updates to happen, see
+      # # https://www.postgresql.org/docs/9.4/explicit-locking.html
+      source = Keyword.get(opts, :migration_source, "schema_migrations")
+      table = if prefix = opts[:prefix], do: ~s|"#{prefix}"."#{source}"|, else: ~s|"#{source}"|
+      {:ok, _} = Ecto.Adapters.SQL.query(meta, "LOCK TABLE #{table} IN SHARE UPDATE EXCLUSIVE MODE", [], opts)
+      fun.()
+    end)
```

Comment:
```suggestion
    advisory_lock(meta, opts, lock, fun)
  end

  defp do_lock_for_migrations(_lock_strategy, meta, opts, fun) do
    {:ok, res} =
      transaction(meta, opts, fn ->
        # SHARE UPDATE EXCLUSIVE MODE is the first lock that locks
        # itself but still allows updates to happen, see
        # # https://www.postgresql.org/docs/9.4/explicit-locking.html
        source = Keyword.get(opts, :migration_source, "schema_migrations")
        table = if prefix = opts[:prefix], do: ~s|"#{prefix}"."#{source}"|, else: ~s|"#{source}"|
        {:ok, _} = Ecto.Adapters.SQL.query(meta, "LOCK TABLE #{table} IN SHARE UPDATE EXCLUSIVE MODE", [], opts)
        fun.()
      end)
    
    res
```

This way you don't need `{:ok, result} = do_lock_for_migrations(lock_strategy, meta, opts, fun)`, you can return  `do_lock_for_migrations(lock_strategy, meta, opts, fun)` directly.